File: queues/queue_skeleton.cpp
// Practice 3. Queue
#include <cstdlib>
#include <iostream>
#include <fstream>
#include <sstream>
using namespace std;
const char ENQUEUE = 'E';
const char DEQUEUE = 'D';
const char FRONT = 'F';
const char PUSH = 'U';
const char POP = 'O';
const char TOP = 'T';

struct Node
{
  int element;
  Node *next;
};

class Queue
{
private:
  Node *head;

public:
  Queue()
  {
    head = nullptr;
  }
  ~Queue()
  {
    Node *ptr = head;
    while (ptr != nullptr)
    {
      Node *temp = ptr->next;
      delete ptr;
      ptr = temp;
    }
  }

  // Enqueue an ele to the queue
  bool enqueue(int d)
  {
    Node *temp = new Node;
    temp->element = d;
    temp->next = nullptr;

    if (head == nullptr)
    {
      head = temp;
    }
    else
    {
      Node *ptr = head;
      while (ptr->next != nullptr)
      {
        ptr = ptr->next;
      }
      ptr->next = temp;
    }

    return 1;
  }

  // Dequeue an ele from the queue and return the its value
  int dequeue()
  {
    if (head == nullptr)
    {
      return 0;
    }
    else
    {
      Node *ptr = head;
      head = ptr->next;
      int temp = ptr->element;
      delete ptr;

      return temp;
    }
  }

  // Get the front of the queue
  int front()
  {
    return head->element;
  }

  // Write the value of every ele in the queue to the output file
  void write(ofstream &outFile)
  {
    Node *ptr = head;
    while (ptr != nullptr)
    {
      outFile << ptr->element << " ";

      ptr = ptr->next;
    }
    outFile << endl;
  }
};

class StackViaQueues
{
private:
  Queue q1;
  Queue q2;
  int len = 0;

public:
  // TODO. Define constructor/destructor/methods here
  // Push an ele to the stack
  bool push(int d)
  {
    q2.enqueue(d);

    for (int i = 0; i < len; i++)
    {
      q2.enqueue(q1.dequeue());
    }

    Queue temp = q1;
    q1 = q2;
    q2 = temp;
    len++;

    return 1;
  }

  // Pop the top ele from the stack
  int pop()
  {
    int re = q1.dequeue();
    len--;

    return re;
  }

  // Retrieve the top of the stack
  int peek()
  {
    return q1.front();
  }

  // Write the value of every ele from the bottom to the top
  // to the output file
  void write(ofstream &outFile)
  {
    int arr[len];
    for (int i = 0; i < len; i++)
    {
      arr[i] = q1.dequeue();
      q1.enqueue(arr[i]);
    }
    for (int i = len - 1; i >= 0; i--)
    {
      outFile << arr[i] << " ";
    }
    outFile << endl;
  }
};

int main(int argc, char *argv[])
{
  if (argc != 3)
  {
    cerr << "Correct usage: [program] [input] [output]" << endl;
    exit(1);
  }
  Queue q;
  StackViaQueues s;
  ifstream inFile(argv[1]);
  ofstream outFile(argv[2]);
  string line;
  while (getline(inFile, line))
  {
    char op = line[0];
    istringstream iss(line.substr(1));
    int data;
    switch (op)
    {
    case ENQUEUE:
      if (!(iss >> data))
      {
        cerr << "ADD: invalid input" << endl;
        exit(1);
      }
      if (q.enqueue(data))
        q.write(outFile);
      else
        outFile << "Enqueue failed" << endl;
      break;
    case DEQUEUE:
      q.dequeue();
      q.write(outFile);
      break;
    case FRONT:
      outFile << q.front() << endl;
      break;
    case PUSH:
      if (!(iss >> data))
      {
        cerr << "PUSH: invalid input" << endl;
        exit(1);
      }
      if (s.push(data))
        s.write(outFile);
      else
        outFile << "Push failed" << endl;
      break;
    case POP:
      s.pop();
      s.write(outFile);
      break;
    case TOP:
      data = s.peek();
      outFile << data << endl;
      break;
    default:
      cerr << "Undefined operator" << endl;
      exit(1);
    }
  }
  outFile.close();
  inFile.close();
}

File: queues/queue_skeleton.py
# Practice 3. Queue
import sys

class Empty(Exception):
  pass

class Queue:
  def __init__(self):
    # TODO
    pass

  # Enqueue an element to the queue
  def enqueue(self, d):
    # TODO
    pass

  # Dequeue an element from the queue and return the its value
  def dequeue(self):
    # TODO
    pass

  # Get the front of the queue
  def front(self):
    # TODO
    pass

  # Write the value of every element in the queue to the output file
  def write(self, outFile):
    # TODO
    pass

# Stack via queues
class StackViaQueues:
  def __init__(self):
    # TODO
    pass

  # Push an element to the stack
  def push(self, d):
    # TODO
    pass

  # Pop the top element from the stack
  def pop(self):
    # TODO
    pass

  # Retreive the top of the stack
  def peek(self):
    # TODO
    pass

  # Write the value of every element from the bottom to the top 
  # to the output file
  def write(self, outFile):
    # TODO
    pass

if __name__ == "__main__":
  if len(sys.argv) != 3:
    raise Exception("Correct usage: [program] [input] [output]")
  Q = Queue()
  S = StackViaQueues()
  with open(sys.argv[1], 'r') as inFile:
    lines = inFile.readlines()
  with open(sys.argv[2], 'w') as outFile:
    for line in lines:
      words = line.split()
      op = words[0]
      if op == 'E':
        if len(words) != 2:
          raise Exception("ENQUEUE: invalid input")
        if Q.enqueue(words[1]):
          Q.write(outFile)
        else:
          outFile.write("Enqueue failed\n")   
      elif op == 'D':
        if len(words) != 1:
          raise Exception("DEQUEUE: invalid input")
        Q.dequeue()
        Q.write(outFile)
      elif op == 'F':
        if len(words) != 1:
          raise Exception("FRONT: invalid input")
        val = Q.front()
        outFile.write(val + "\n")
      elif op == 'U':
        if len(words) != 2:
          raise Exception("PUSH: invalid input")
        S.push(words[1])
        S.write(outFile)
      elif op == 'O':
        if len(words) != 1:
          raise Exception("POP: invalid input")
        S.pop()
        S.write(outFile)
      elif op == 'T':
        if len(words) != 1:
          raise Exception("TOP: invalid input")
        val = S.peek()
        outFile.write(val + "\n")
      else:
        raise Exception("Undefined operator")

